directory/directory_dest.go

```diff
@@ -100,7 +100,7 @@ func (d *dirImageDestination) SupportsSignatures(ctx context.Context) error {
 	return nil
 }
 
-func (d *dirImageDestination) DesiredLayerCompression() types.LayerCompression {
+func (d *dirImageDestination) DesiredLayerCompression(info types.BlobInfo) types.LayerCompression {
```

Comment:
Ok, I just updated it per the above.

Reply:
It’s worse than that: callers of e.g. `copy.Image` can, in principle, submit their own `ImageReference` implementation that returns their own `ImageDestination` implementation, so adding any new methods to `ImageDestination` is also an API-breaking change. (And there is at least Buildah’s blob caching that actually does provide its own implementation/wrapper.)

We’ve been talking about the need to make `ImageDestination` and several other types (mostly) private for some time now; between this, the delta support, and enhancing `BlobInfoCache`, we’ll need to figure this out now.

I’ll read through the various PRs and try to figure out a way.

Reply:
… but, looking at this PR more, this might not be quite the concern for _this_ one.

Conceptually, whether a MIME type can be compressed is a property of either that MIME type _only_, or of a combination of that MIME type and the image format ~ manifest schema (`types.Image` implementation); AFAICS is not really the property of the image transport (e.g. `dir:` and `containers-storage:` can accept/supply several different image formats).

So I think this should conceptually be a new method on `types.Image`, not on `ImageDestination`, unless there’s something that would make that impractical.

That _still_ has the concern of breaking API by adding extra methods to `types.Image`, but the only implementation the copy pipeline cares about is the one returned by `image.FromUnparsedImage`, so we don’t have to add anything to the public API. (The `ImageReference.NewImage` method is somewhat a historical artifact.)

Tentatively I’ve been thinking:
- Define an `c/image/internal/image.Image`, which embeds `c/image/types.Image` but adds the necessary new method.
- Move the implementation of `c/image.sourcedImage` to `c/image/internal/image`; add the necessary new method there. Make the existing public API in `c/image/sourced.go` compatibility wrappers for the internal versions.
- (From a quick look it seems that the actual implementation would be a added to the various c/manifest types, which are not interfaces, so we can add methods safely.)

… but I still need to examine the other PRs, or try doing this myself — there may well be something about this plan that doesn’t work.

Reply:
So it seems like maybe the best path forward is the original patch?

Reply:
The original patch still not solving the problem directly.

c/image needs to deal with the API anyway. It’s completely fair that you shouldn’t have to sign up for that work just to get this fix in — I’m going to be working in this area now, but I can’t promise that it will be done by any specific date.